### DISCUSSION TITLE
ServiceStack.RabbitMQ compatibility with .NET Aspire

### DISCUSSION THREAD

**Discussion by Ryan-Britton** [ServiceStack/Discuss#102] [Q&A] [answered]
I'm creating a solution which uses the new [.NET Aspire](https://learn.microsoft.com/en-us/dotnet/aspire/get-started/aspire-overview) orchestrator support. I would (obviously) love to be able to use the Aspire components for RabbitMQ and Redis along with the ServiceStack support for the same technologies (see [here](https://learn.microsoft.com/en-us/dotnet/aspire/messaging/rabbitmq-client-component?tabs=dotnet-cli) and [here](https://learn.microsoft.com/en-us/dotnet/aspire/caching/stackexchange-redis-component?tabs=dotnet-cli)) - but I am struggling to get the RabbitMQ component to play nicely with the ServiceStack RabbitMQ implementation. 

In theory, it sounds like this should be achievable - both rely on the RabbitMQ .NET client, and I can inject a wired-up RabbitMQ ConnectionFactory into a ServiceStack RabbitMqConnectionFactory instance when setting up the MQ Server:

```
public class ConfigureMq : IHostingStartup
    {
        public void Configure(IWebHostBuilder builder) => builder
            .ConfigureServices((context, services) =>
            {
                
                services.AddSingleton<IMessageService>(c => 
                    new RabbitMqServer(new RabbitMqMessageFactory(c.GetRequiredService<IConnectionFactory>() as ConnectionFactory)));
            })
            .ConfigureAppHost(afterAppHostInit: appHost =>
            {
                var msgService = appHost.Resolve<IMessageService>();
                msgService.RegisterHandler<MqExecutePing>(appHost.ServiceController.ExecuteMessage);
                appHost.Resolve<IMessageService>().Start();
            });
    }
```

unfortunately, this seems to throw the following error when trying to publish a message : 

```
System.MissingMethodException: Method not found: 'Void RabbitMQ.Client.IModelExensions.BasicPublish(RabbitMQ.Client.IModel, System.String, System.String, RabbitMQ.Client.IBasicProperties, Byte[])'.
   at ServiceStack.RabbitMq.RabbitMqProducer.PublishMessage(String exchange, String routingKey, IBasicProperties basicProperties, Byte[] body)
   at ServiceStack.RabbitMq.RabbitMqProducer.Publish(String queueName, IMessage message, String exchange) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack.RabbitMq/RabbitMqProducer.cs:line 117
   at ServiceStack.RabbitMq.RabbitMqProducer.Publish(String queueName, IMessage message) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack.RabbitMq/RabbitMqProducer.cs:line 76
   at ServiceStack.RabbitMq.RabbitMqProducer.Publish[T](T messageBody) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack.RabbitMq/RabbitMqProducer.cs:line 61
   at ServiceStack.RabbitMq.RabbitMqProducer.Publish[T](T messageBody) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack.RabbitMq/RabbitMqProducer.cs:line 65
   at Redacted.Api.Accounts.ServiceInterface.HealthServices.Any(MqPing request) in D:\repos\Redacted\src\Redacted.Api.Accounts\ServiceInterface\HealthServices.cs:line 16
   at lambda_method2(Closure, Object, Object)
   at ServiceStack.Host.ServiceRunner`1.ExecuteAsync(IRequest req, Object instance, TRequest requestDto) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack/Host/ServiceRunner.cs:line 131
```

Could this be because of version differences between the underlying RabbitMQ .NET client? Is there a way around this?

**Comment by mythz** [comment]
It's because it's using a different / incompatible version [RabbitMQ.Client](https://www.nuget.org/packages/RabbitMQ.Client/), we haven't upgraded since it contains several breaking changes, but I'll move this to a feature request to upgrading to the latest version for later.

**Comment by Ryan-Britton** [comment]
[PR](https://github.com/ServiceStack/ServiceStack/pull/1339) submitted

**Comment by Ryan-Britton** [comment]
Looks like there is still an issue - I get the following error:

```
System.MissingMethodException: Method not found: 'Void RabbitMQ.Client.IModelExensions.BasicPublish(RabbitMQ.Client.IModel, System.String, System.String, RabbitMQ.Client.IBasicProperties, Byte[])'.
   at ServiceStack.RabbitMq.RabbitMqProducer.PublishMessage(String exchange, String routingKey, IBasicProperties basicProperties, Byte[] body)
   at ServiceStack.RabbitMq.RabbitMqProducer.Publish(String queueName, IMessage message, String exchange) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack.RabbitMq/RabbitMqProducer.cs:line 117
   at ServiceStack.RabbitMq.RabbitMqProducer.Publish(String queueName, IMessage message) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack.RabbitMq/RabbitMqProducer.cs:line 76
   at ServiceStack.RabbitMq.RabbitMqProducer.Publish[T](T messageBody) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack.RabbitMq/RabbitMqProducer.cs:line 61
   at ServiceStack.RabbitMq.RabbitMqProducer.Publish[T](T messageBody) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack.RabbitMq/RabbitMqProducer.cs:line 65
   at Redacted.Api.Accounts.ServiceInterface.HealthServices.Any(MqPing request) in D:\repos\Redacted\src\Redacted.Api.Accounts\ServiceInterface\HealthServices.cs:line 16
   at lambda_method2(Closure, Object, Object)
   at ServiceStack.Host.ServiceRunner`1.ExecuteAsync(IRequest req, Object instance, TRequest requestDto) in /home/runner/work/ServiceStack/ServiceStack/ServiceStack/src/ServiceStack/Host/ServiceRunner.cs:line 131
```

looking into the code - it looks like there is an implicit, conversion from byte[] to ReadOnlyMemory<byte> which is failing when calling Channel.BasicPublish (the implicit conversion is obviously why there is no build error) : 

```
public virtual void PublishMessage(string exchange, string routingKey, IBasicProperties basicProperties, byte[] body)
        {
            try
            {
                // In case of server named queues (client declared queue with channel.declare()), assume queue already exists
                //(redeclaration would result in error anyway since queue was marked as exclusive) and publish to default exchange
                if (routingKey.IsServerNamedQueue())
                {
                    Channel.BasicPublish("", routingKey, basicProperties, body);
                }
                else
                {
                    if (!Queues.Contains(routingKey))
                    {
                        Channel.RegisterQueueByName(routingKey);
                        Queues = new HashSet<string>(Queues) { routingKey };
                    }

                    Channel.BasicPublish(exchange, routingKey, basicProperties, body);
                }

            }
            catch (OperationInterruptedException ex)
            {
                if (ex.Is404())
                {
                    // In case of server named queues (client declared queue with channel.declare()), assume queue already exists (redeclaration would result in error anyway since queue was marked as exclusive) and publish to default exchange
                    if (routingKey.IsServerNamedQueue())
                    {
                        Channel.BasicPublish("", routingKey, basicProperties, body);
                    }
                    else
                    {
                        Channel.RegisterExchangeByName(exchange);

                        Channel.BasicPublish(exchange, routingKey, basicProperties, body);
                    }
                }
                throw;
            }
        }
```

here is the method called in the RabbitMq.Client library:

```
/// <summary>
/// (Extension method) Convenience overload of BasicPublish.
/// </summary>
/// <remarks>
/// The publication occurs with mandatory=false
/// </remarks>
public static void BasicPublish(this IModel model, string exchange, string routingKey, IBasicProperties basicProperties, ReadOnlyMemory<byte> body)
{
    model.BasicPublish(exchange, routingKey, false, basicProperties, body);
}
```

I'm not sure if it's appropriate to just do a cast here, or if there is a better way of handling this?

**Reply by mythz** [reply]
> looking into the code - it looks like there is an implicit, conversion from byte[] to ReadOnlyMemory which is failing when calling Channel.BasicPublish

The issue isn't the implicit conversion of `byte[]` with `ReadOnlyMemory<byte>` which is a safe conversion.

`MissingMethodException` suggests that you're App is using a different RabbitMQ version with an incompatible method signature than what ServiceStack was built against.

I've just published these changes to the v8.0.1 [Pre Release packages](https://docs.servicestack.net/pre-release) which you can use in Apps that use the latest RabbitMQ.Client v6.8.1.

**Reply by Ryan-Britton** [reply]
it's so strange - if I grab the RabbitMQ.Client reference properties from the project it seems to be pointing to the 6.8.1 RabbitMq.Client version : 

![image](https://github.com/ServiceStack/Discuss/assets/124330161/df8cbc64-0ad4-4eee-8f6e-24b4fb01a804)

I cleared my nuget cache (just in case) using the method described [here](https://docs.servicestack.net/pre-release#clear-nuget-package-cache) : and explicitly installed the RabbitMq.Client into the project : 

![image](https://github.com/ServiceStack/Discuss/assets/124330161/ac729e6e-8026-4885-8bb2-03c69d7c7b85)


no dice - I get the same error...

**Reply by mythz** [reply]
Weird, If you put together a stand-alone repro on GitHub I can take a look.

**Reply by Ryan-Britton** [reply, marked as answer]
just wanted to close this loop : I created a small repo and a ServiceStack solution by running : 

- x new web ExampleApp
- x mix feature-mq rabbitmq feedz
- upgrading the ServiceStack packages to v8.0.1
- adding SharpPagesFeature manually (itwas missing from the feature-mq files and I couldn't find a mix for it)
- successfully publishing to the latest rabbitmq-management image running in docker

for all intents and purposes, it looks like whatever issue I am having is not related to the baseline upgrade

The repo is here if you want to take a look : https://github.com/Ryan-Britton/ServiceStack-RabbitMQUpgradeExample

**Comment by Misiu** [comment]
I'll be taking over an old ServiceStack (.NET5, ServiceStack 5.13.2) to the latest version of .NET and ServiceStack.
I want to use Aspire as it simplifies the development process, especially if your solution requires many dependant services like redis, rabbitmq, database etc.
Is Aspire officially supported? Is there an example showing how to use it with ServiceStack? Sadly, @Ryan-Britton repo isn't available anymore.

**Reply by mythz** [reply]
ServiceStack doesn't have any dependencies on Aspire, but from [Aspire's RabbitMQ Docs](https://learn.microsoft.com/en-us/dotnet/aspire/messaging/rabbitmq-integration?tabs=dotnet-cli) you should be able to configure it directly to use a RabbitMQ connection.

**Reply by Misiu** [reply]
@mythz thank you for an update. I'll try to create a POC in the following days. The idea is to use Aspire to set the development environment, to help new developers set up the environment. Right now, there are a lot of things that must be set manually, and I'd like to use Aspire for that.
